File: hbTapi Components/Delphi Samples/Enterprise Edition/FaxHandoff/uMain.pas
{-------------------------------------------------------------------------
   Copyright by Haeger + Busch, Germany                / >>>>>>>>> /-----
   Ingenieurbuero fuer Kommunikationslösungen         / <<<<<<<<< /
 ----------------------------------------------------/ >>>>>>>>> /
   Homepage  : http://www.hbTapi.com
   EMail     : info@hbTapi.com
   Package   : hbTapi Components
 -------------------------------------------------------------------------}
unit uMain;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  StdCtrls, hbTAPI, hbTapiApi, ExtCtrls;

type
  TForm1 = class(TForm)
    CheckBox_Active: TCheckBox;
    ComboBoxDevice: TComboBox;
    EditApplication: TEdit;
    Button_Answer: TButton;
    Button_Drop: TButton;
    LabelApplication: TLabel;
    hbTapiLine1: ThbTapiLine;
    ListBox1: TListBox;
    procedure CheckBox_ActiveClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure ComboBoxDeviceChange(Sender: TObject);
    procedure Button_AnswerClick(Sender: TObject);
    procedure Button_DropClick(Sender: TObject);
    procedure ButtonClearClick(Sender: TObject);
    procedure hbTapiLine1MediaMode(Sender: ThbTapiLine; Call: ThbTapiCall);
    procedure hbTapiLine1Connected(Sender: ThbTapiLine; Call: ThbTapiCall);
    procedure hbTapiLine1Ring(Sender: ThbTapiLine; RingMode,
      RingCount: Cardinal);
    procedure hbTapiLine1CallState(Sender: ThbTapiLine; Call: ThbTapiCall;
      CallState: Cardinal);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  Form1: TForm1;

implementation

{$R *.DFM}

procedure TForm1.CheckBox_ActiveClick(Sender: TObject);
begin
  if not CheckBox_Active.Checked then
    begin
      if hbTapiLine1.Active then
        begin
          ListBox1.Items.Add('hbTapiLine1.Active = False');
          try
            hbTapiLine1.Active := False;
          except
            CheckBox_Active.Checked := True;
            raise;
          end;
        end;
    end
  else
  begin
    if not hbTapiLine1.Active then
    begin
      ListBox1.Items.Add('hbTapiLine1.Active = True');
      try
        hbTapiLine1.Active := True;
      except
        on E:EhbTapiError do
        begin
          if E.ErrorCode = LINEERR_INVALMEDIAMODE then
          begin
            ListBox1.Items.Add('This device is not 100% TAPI compliant');
            hbTapiLine1.MediaModes := LINEMEDIAMODE_DATAMODEM;
            try
              hbTapiLine1.Active := True;
            except
              ListBox1.Items.Add('Error = ' + E.Message);
              CheckBox_Active.Checked := False;
            end;
          end
          else
            CheckBox_Active.Checked := False;
        end;
      end;
    end;
  end;
end;

{------------------------------------------------------------------------------}
{ FormCreate
{------------------------------------------------------------------------------}
procedure TForm1.FormCreate(Sender: TObject);
begin
  ComboBoxDevice.Items := hbTapiLine1.DeviceList;
  if ComboBoxDevice.Items.Count > 0 then
    ComboBoxDevice.Text := ComboBoxDevice.Items[0];
  ComboBoxDevice.ItemIndex := 0;
end;

{------------------------------------------------------------------------------}
{ ComboBoxDeviceChange
{------------------------------------------------------------------------------}
procedure TForm1.ComboBoxDeviceChange(Sender: TObject);
begin
  if hbTapiLine1.Active then
    CheckBox_Active.Checked := False;
  ListBox1.Items.Add('hbTapiLine1.DeviceID = ' + IntToStr(ComboBoxDevice.ItemIndex));
  hbTapiLine1.DeviceID := ComboBoxDevice.ItemIndex;
  if not hbTapiLine1.Available then
    ListBox1.Items.Add('Device not available')
  else
  begin
    hbTapiLine1.Privileges.Monitor := True;
    hbTapiLine1.Privileges.Owner := True;
  end;
end;

{------------------------------------------------------------------------------}
{ ButtonAnswerClick
{------------------------------------------------------------------------------}
procedure TForm1.Button_AnswerClick(Sender: TObject);
begin
  ListBox1.Items.Add('Answering');
  if hbTapiLine1.Calls.Count > 0 then
  begin
    try
      hbTapiLine1.Calls[0].Answer;
    except
      on E:Exception do
        ListBox1.Items.Add(#9 + 'Error = ' + E.Message);
    end;
  end
  else
    ListBox1.Items.Add(#9 + 'No Call');
end;

{------------------------------------------------------------------------------}
{ ButtonHangupClick
{------------------------------------------------------------------------------}
procedure TForm1.Button_DropClick(Sender: TObject);
begin
  ListBox1.Items.Add('Drop');
  if hbTapiLine1.Calls.Count > 0 then
  begin
    try
      hbTapiLine1.Calls[0].Drop;
    except
      on E:Exception do
        ListBox1.Items.Add('Error = ' + E.Message);
    end;
  end
  else
    ListBox1.Items.Add('No Call');
end;

{------------------------------------------------------------------------------}
{ ButtonClearClick
{------------------------------------------------------------------------------}
procedure TForm1.ButtonClearClick(Sender: TObject);
begin
  ListBox1.Items.Clear;
end;

procedure TForm1.hbTapiLine1MediaMode(Sender: ThbTapiLine; Call: ThbTapiCall);
begin
  ListBox1.Items.Add('ThbTapiLine.OnMediaMode');
  if (Call.MediaMode and LINEMEDIAMODE_G3FAX) = 0 then
    ListBox1.Items.Add(#9 + 'MediaMode <> LINEMEDIAMODE_G3FAX')
  else
    begin
        ListBox1.Items.Add(#9 + 'MediaMode = LINEMEDIAMODE_G3FAX');
        ListBox1.Items.Add('Call.HandOff(' + #39 + EditApplication.Text + #39 + ')');
        try
          Call.HandOff(EditApplication.Text);
        except
          on E:Exception do
            ListBox1.Items.Add(#9 + 'Error = ' + E.Message);
        end;
    end;
end;

procedure TForm1.hbTapiLine1Connected(Sender: ThbTapiLine; Call: ThbTapiCall);
begin
  ListBox1.Items.Add('::OnConnected');
  if (Call.Features and LINECALLFEATURE_MONITORMEDIA) = 0 then
  begin
    ListBox1.Items.Add('MonitorMedia not supported. We will try it anyway.');
  end;
  try
    ListBox1.Items.Add('Call.MonitorMedia.MediaModes := LINEMEDIAMODE_G3FAX');
    ListBox1.Items.Add('Call.MonitorMedia.Active := True');
    Call.MonitorMedia.MediaModes := LINEMEDIAMODE_G3FAX;
    Call.MonitorMedia.Active := True;
  except
    on E:Exception do
      ListBox1.Items.Add('Error = ' + E.Message)
  end;
end;

procedure TForm1.hbTapiLine1Ring(Sender: ThbTapiLine; RingMode,
  RingCount: Cardinal);
begin
  ListBox1.Items.Add('hbTapiLine1.OnRing #' + IntToStr(RingCount));
end;

procedure TForm1.hbTapiLine1CallState(Sender: ThbTapiLine; Call: ThbTapiCall;
  CallState: Cardinal);
begin
  ListBox1.Items.Add(Call.StateText);
end;

end.

File: hbTapi Components/BCB Samples/Enterprise Edition/FaxHandoff/uMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "uMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "hbTapi"
#pragma resource "*.dfm"
TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
        : TForm(Owner)
{
}
//---------------------------------------------------------------------------


void __fastcall TForm1::FormCreate(TObject *Sender)
{
  hbTapiLine1->Privileges->AutoSelect = false;
  hbTapiLine1->Privileges->Monitor = true;
  hbTapiLine1->Privileges->Owner = true;
  ComboBoxDevice->Items->Assign(hbTapiLine1->DeviceList);
  if (ComboBoxDevice->Items->Count > 0 )
    ComboBoxDevice->ItemIndex = 0;
}
//---------------------------------------------------------------------------

void __fastcall TForm1::CheckBoxActiveClick(TObject *Sender)
{
  if (!CheckBoxActive->Checked)
  {
    if (hbTapiLine1->Active )
    {
      hbTapiLine1->Active = false;
      ListBox1->Items->Add("hbTapiLine1->Active = false");
    }
  }
  else
  {
    if (!hbTapiLine1->Active )
    {
      hbTapiLine1->Privileges->AutoSelect = false;
      hbTapiLine1->Privileges->Owner = true; // we want to be the owner of inbound calls
      hbTapiLine1->Privileges->Monitor = true;  // we want to see all calls
      hbTapiLine1->MediaModes = 0; // all supported media modes
      try
      {
        hbTapiLine1->Active = true;
      }
      catch (EhbTapiError &E)
      {
        if (E.ErrorCode == LINEERR_INVALMEDIAMODE)
        {
          hbTapiLine1->MediaModes = LINEMEDIAMODE_DATAMODEM;
          hbTapiLine1->Active = true;
        }
      }
      ListBox1->Items->Add("hbTapiLine1->Active = true");
    };
  };
  CheckBoxActive->Checked = hbTapiLine1->Active;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::ComboBoxDeviceChange(TObject *Sender)
{
  if (hbTapiLine1->Active)
  {
    CheckBoxActive->Checked = false;
    hbTapiLine1->Active = false;
    ListBox1->Items->Add("hbTapiLine1->Active = false");
  }

  ListBox1->Items->Add("hbTapiLine1->DeviceID = " + IntToStr(ComboBoxDevice->ItemIndex));
  hbTapiLine1->DeviceID = ComboBoxDevice->ItemIndex;

  if (!hbTapiLine1->Available)
    ListBox1->Items->Add("Device not available!");
}
//---------------------------------------------------------------------------
void __fastcall TForm1::ButtonAnswerClick(TObject *Sender)
{
  ListBox1->Items->Add("Call->Answering()");
  if (hbTapiLine1->Calls->Count > 0 )
  {
    try
    {
      hbTapiLine1->Calls->Items[0]->Answer("");
    }
    catch (Exception &E)
    {
      ListBox1->Items->Add("Error = " + E.Message);
    };
  }
  else
    ListBox1->Items->Add("No Call");

}
//---------------------------------------------------------------------------
void __fastcall TForm1::ButtonHangupClick(TObject *Sender)
{
  ListBox1->Items->Add("Call->Drop()");
  if (hbTapiLine1->Calls->Count > 0 )
  {
    try
    {
      hbTapiLine1->Calls->Items[0]->Drop("");
    }
    catch (Exception &E)
    {
      ListBox1->Items->Add("Error = " + E.Message);
    };
  }
  else
    ListBox1->Items->Add("No Call");
}

//---------------------------------------------------------------------------
void __fastcall TForm1::hbTapiLine1Connected(ThbTapiLine *Sender,
      ThbTapiCall *Call)
{
  ListBox1->Items->Add("::OnConnected");
  if ((Call->Features & LINECALLFEATURE_MONITORMEDIA) == 0)
  {
    ListBox1->Items->Add("MonitorMedia is not supported. We try it anyway.");
  }
  try
  {
    ListBox1->Items->Add("Call->MonitorMedia->MediaModes = LINEMEDIAMODE_G3FAX");
    ListBox1->Items->Add("Call->MonitorMedia->Active = true");
    Call->MonitorMedia->MediaModes = LINEMEDIAMODE_G3FAX;
    Call->MonitorMedia->Active = true;
  }
  catch (Exception &E)
  {
    ListBox1->Items->Add("Error = " + E.Message);
  };
}
//---------------------------------------------------------------------------
void __fastcall TForm1::hbTapiLine1MediaMode(ThbTapiLine *Sender,
      ThbTapiCall *Call)
{
  ListBox1->Items->Add("ThbTapiLine->OnMediaMode");
  if ((Call->MediaMode & LINEMEDIAMODE_G3FAX) == 0)
    ListBox1->Items->Add("MediaMode != LINEMEDIAMODE_G3FAX");
  else
  {
    ListBox1->Items->Add("MediaMode == LINEMEDIAMODE_G3FAX");
    ListBox1->Items->Add("Call->HandOff(EditApplication->Text)");
    try
    {
      Call->HandOff(EditApplication->Text);
    }
    catch (Exception &E)
    {
      ListBox1->Items->Add("Error = " + E.Message);
    };
  }
}
//---------------------------------------------------------------------------
void __fastcall TForm1::hbTapiLine1CallState(ThbTapiLine *Sender,
      ThbTapiCall *Call, DWORD CallState)
{
  ListBox1->Items->Add("Call->State == " + Call->StateText());
}
//---------------------------------------------------------------------------
